File: crc32c.py
"""
CRC32C (Castagnoli) implementation with a precomputed table.
Pure Python fallback to avoid external dependencies.
"""

_POLY = 0x1EDC6F41


def _make_table():
    tbl = []
    for n in range(256):
        c = n
        for _ in range(8):
            if c & 1:
                c = (c >> 1) ^ _POLY
            else:
                c >>= 1
        tbl.append(c & 0xFFFFFFFF)
    return tuple(tbl)


_TABLE = _make_table()


def crc32c(data: bytes, crc: int = 0) -> int:
    c = (~crc) & 0xFFFFFFFF
    for b in data:
        c = _TABLE[(c ^ b) & 0xFF] ^ (c >> 8)
    return (~c) & 0xFFFFFFFF

File: src/crc32c.rs
const POLY: u32 = 0x1EDC6F41;

const fn make_table() -> [u32; 256] {
    let mut tbl = [0u32; 256];
    let mut n = 0usize;
    while n < 256 {
        let mut c = n as u32;
        let mut bit = 0;
        while bit < 8 {
            if (c & 1) != 0 {
                c = (c >> 1) ^ POLY;
            } else {
                c >>= 1;
            }
            bit += 1;
        }
        tbl[n] = c;
        n += 1;
    }
    tbl
}

const TABLE: [u32; 256] = make_table();

pub fn crc32c(data: &[u8], crc: u32) -> u32 {
    let mut c = !crc;
    for &byte in data {
        c = TABLE[((c ^ (byte as u32)) & 0xFF) as usize] ^ (c >> 8);
    }
    !c
}

#[cfg(test)]
#[path = "tests/crc32c.rs"]
mod tests;
